### PR TITLE
Specialize `PartialOrd<A> for [A] where A: Ord`

### PR DESCRIPTION
**Pull request by ghost**
This way we can call `cmp` instead of `partial_cmp` in the loop, removing some burden of optimizing `Option`s away from the compiler.

PR #39538 introduced a regression where sorting slices suddenly became slower, since `slice1.lt(slice2)` was much slower than `slice1.cmp(slice2) == Less`. This problem is now fixed.

To verify, I benchmarked this simple program:
```rust
fn main() {
    let mut v = (0..2_000_000).map(|x| x * x * x * 18913515181).map(|x| vec![x, x ^ 3137831591]).collect::<Vec<_>>();
    v.sort();
}
```

Before this PR, it would take 0.95 sec, and now it takes 0.58 sec.
I also tried changing the `is_less` lambda to use `cmp` and `partial_cmp`. Now all three versions (`lt`, `cmp`, `partial_cmp`) are equally performant for sorting slices - all of them take 0.58 sec on the
benchmark.

Tangentially, as soon as we get `default impl`, it might be a good idea to implement a blanket default impl for `lt`, `gt`, `le`, `ge` in terms of `cmp` whenever possible. Today, those four functions by default are only implemented in terms of `partial_cmp`.

r? @alexcrichton

### CODE REVIEW

File: src/libcore/slice.rs [hunk 1]
```diff
@@ -2290,6 +2290,28 @@ impl<A> SlicePartialOrd<A> for [A]
...
+                Ordering::Equal => (),
+                non_eq => return Some(non_eq),
+            }
+        }
+
+        self.len().partial_cmp(&other.len())
+    }
+}
+
 impl SlicePartialOrd<u8> for [u8] {
```
Comment (ollie27) [inline src/libcore/slice.rs]: Could this impl not be extended to all `A: Ord` so it would simply use `Some(SliceOrd::compare(self, other))`?

Comment (stjepang) [inline src/libcore/slice.rs]: That's a good suggestion, thanks! I've updated the code.

File: src/libcore/slice.rs [hunk 2]
```diff
@@ -2290,6 +2290,14 @@ impl<A> SlicePartialOrd<A> for [A]
...
 
+impl<A> SlicePartialOrd<A> for [A]
+    where A: Ord
+{
+    default fn partial_compare(&self, other: &[A]) -> Option<Ordering> {
+        Some(SliceOrd::compare(self, other))
+    }
+}
+
 impl SlicePartialOrd<u8> for [u8] {
```
Comment (alexcrichton) [inline src/libcore/slice.rs]: I think this specialization could be removed now, right?

**Comment by alexcrichton** [comment]
Looks good to me, thanks @stjepang!

cc @rust-lang/libs, @bluss, I'm sure this has the likelihood of being super subtle, so would be good to get some more eyes on this as well

File: src/libcore/slice.rs (hunk 2 above)
Comment (stjepang) [inline src/libcore/slice.rs]: Yeah, removed.

**Pull request merged**